datalad/interface/run.py

```diff
@@ -0,0 +1,268 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Run arbitrary commands and track how they modify a dataset"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+import json
+import re
+
+from argparse import REMAINDER
+from os.path import join as opj
+from os.path import curdir
+from os.path import pardir
+from os.path import relpath
+from os.path import normpath
+
+from datalad.cmd import Runner
+
+from datalad.interface.base import Interface
+from datalad.interface.utils import eval_results
+from datalad.interface.base import build_doc
+from datalad.interface.results import get_status_dict
+from datalad.interface.common_opts import save_message_opt
+
+from datalad.support.constraints import EnsureNone
+from datalad.support.constraints import EnsureStr
+from datalad.support.constraints import EnsureChoice
+from datalad.support.exceptions import CommandError
+from datalad.support.exceptions import InsufficientArgumentsError
+from datalad.support.param import Parameter
+
+from datalad.distribution.dataset import Dataset
+from datalad.distribution.dataset import require_dataset
+from datalad.distribution.dataset import EnsureDataset
+from datalad.distribution.dataset import datasetmethod
+
+from datalad.utils import get_dataset_root
+from datalad.utils import getpwd
+from datalad.tests.utils import ok_clean_git
+
+lgr = logging.getLogger('datalad.interface.run')
+
+
+@build_doc
+class Run(Interface):
+    """Run and explain
+
+    If the executed command did not alter the dataset in any way, no record of
+    the command execution is made.
+
+    If the given command errors, a `CommandError` exception with the same exit
+    code will be raised.
+
+    The executed command, the process working directory (PWD; if it is inside
+    the given dataset), and the command's exit code will be recorded in the
+    commit message.
+    """
+    _params_ = dict(
+        cmd=Parameter(
+            args=("cmd",),
+            nargs=REMAINDER,
+            metavar='SHELL COMMAND',
+            doc="command for execution"),
```

Comment:
eventually (now?) we might want to add `--automagicio` (or smth like that) mode which would just run the command (ATM would work only for python scripts) through out automagicIO support (`python -m datalad`).  Sure thing would be applicable only when the command/script is a python script, but imho would improve utility.  Eventually #1739 it should automagically unlock depending on the operation.  And may be even further down the line could run any process while monitoring syscalls....